### DISCUSSION TITLE
How to convert files into vtu with pyvista

### DISCUSSION THREAD

**Discussion by abdelkareemkobo** [pyvista/pyvista#3965] [Q&A] [answered]
I was looking for a way to convert a cube actor into an stl and vtu extensions, and I have managed to convert it to stl but for vtu, things is not clear I coulnd't find a forward code in vtk to do that! i searced the web and found something called vtki and I feel that It's no longer mainted and merged into this pyvista so my question how can i do this simple task with pyvista? 
`
import vtk 
import vtki 
from vtki import examples 
grid = examples.load_hexbeam()
grid.save("grid.vtu")
`
what  is the alternative for code like this in pyvista? thanks in advance

**Comment by banesullivan** [comment, marked as answer]
`vtki`'s name was changed to `pyvista`. Simply change `vtki` to `pyvista`:

```py
import pyvista as pv
from pyvista import examples

grid = examples.load_hexbeam()
grid.save("grid.vtu")
```

**Reply by banesullivan** [reply]
And if for other meshes that may not be an `UnstructuredGrid`. You can use `cast_to_unstructured_grid()`


```py
mesh.cast_to_unstructured_grid().save("grid.vtu")
```

**Reply by abdelkareemkobo** [reply]
thanks, got it. it's strange that pyvista is not located in vtk learning or anything it's wonderful and saves a lot of time and worked

**Reply by abdelkareemkobo** [reply]
so i can read and write with pyvista but I why when I use vtk.vtkPolyDataReader() to read the file I get a lot of errors

```py
import vtk

InputFilename = 'Cube.vtk'
colors = vtk.vtkNamedColors()

#--------------data Source----------
reader = vtk.vtkXMLPolyDataReader()
reader.SetFileName(InputFilename)
reader.Update()

#------------Mapper----------------
mapper = vtk.vtkPolyDataMapper()
mapper.SetInputConnection(reader.GetOutputPort())

#------------Actor----------------
actor = vtk.vtkActor()
actor.SetMapper(mapper)
actor.GetProperty().SetColor(colors.GetColor3d("MistyRose"))

#-----------Renderer & rendering window---------------
renderer = vtk.vtkRenderer()
renderer.AddActor(actor)
renderer.SetBackground(colors.GetColor3d('SlateGray'))

renderWindow = vtk.vtkRenderWindow()
renderWindow.AddRenderer(renderer)
renderWindow.SetSize(500, 500)
renderWindow.Render()

#------------Interactor--------------
iren = vtk.vtkRenderWindowInteractor()
iren.SetRenderWindow(renderWindow)
iren.Start()`
```
the errors I get is : 

```
2023-02-06 19:06:17.346 (   0.152s) [        F9B9C740]       vtkXMLReader.cxx:306    ERR| vtkXMLPolyDataReader (0x55f4530519c0): Error opening file Cube.vtk
2023-02-06 19:06:17.346 (   0.152s) [        F9B9C740]       vtkExecutive.cxx:740    ERR| vtkCompositeDataPipeline (0x55f453052860): Algorithm vtkXMLPolyDataReader (0x55f4530519c0) returned failure for request: vtkInformation (0x55f453054710)
  Debug: Off
  Modified Time: 151
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0


2023-02-06 19:06:17.644 (   0.450s) [        F9B9C740]       vtkXMLReader.cxx:306    ERR| vtkXMLPolyDataReader (0x55f4530519c0): Error opening file Cube.vtk
2023-02-06 19:06:17.644 (   0.450s) [        F9B9C740]       vtkExecutive.cxx:740    ERR| vtkCompositeDataPipeline (0x55f453052860): Algorithm vtkXMLPolyDataReader (0x55f4530519c0) returned failure for request: vtkInformation (0x55f453053270)
  Debug: Off
  Modified Time: 356
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FROM_OUTPUT_PORT: 0
  ALGORITHM_AFTER_FORWARD: 1
  FORWARD_DIRECTION: 0
```

**Reply by banesullivan** [reply]
Are you sure the file "Cube.vtk" exists? Are you sure it is PolyData? If the extension is ".vtk", then try the `vtk.vtkDataSetReader`.

FWIW, just use `pyvista.read()` to read your file. This will choose the correct VTK reader class and still return a VTK object.

**Reply by abdelkareemkobo** [reply]
The file Cube.vtk exists and it's a cube that I created from pyvista with cube mesh. for reading it with pyvista it works fine but the people I will give them the file will read it with normal vtk.vrkXMLPolyDataReader()

**Reply by banesullivan** [reply]
The ".vtk" extension implies this is not an XML PolyData file, but a VTK dataset file, thus requiring the legacy `vtkDataSetReader`. If you must use `vtkXMLPolyDataReader`, then you need to make sure the data are save appropriately. In PyVista, when saving the mesh, use the ".vtp" extension.

```
mesh.save('Cube.vtp')
```

**Reply by abdelkareemkobo** [reply]
I saved the files again with mesh.save("Cube.vtp")  and tried both vtk.vtkXMLReader() and vtk.vtkPolyDataReader() but i get the errors : 
```py
import vtk

InputFilename = 'cube.vtp'
colors = vtk.vtkNamedColors()

#--------------data Source----------
# reader = vtk.vtkXMLPolyDataReader()
reader = vtk.vtkXMLReader()
reader.SetFileName(InputFilename)
reader.Update()

#------------Mapper----------------
mapper = vtk.vtkPolyDataMapper()
mapper.SetInputConnection(reader.GetOutputPort())

#------------Actor----------------
actor = vtk.vtkActor()
actor.SetMapper(mapper)
actor.GetProperty().SetColor(colors.GetColor3d("MistyRose"))

#-----------Renderer & rendering window---------------
renderer = vtk.vtkRenderer()
renderer.AddActor(actor)
renderer.SetBackground(colors.GetColor3d('SlateGray'))

renderWindow = vtk.vtkRenderWindow()
renderWindow.AddRenderer(renderer)
renderWindow.SetSize(500, 500)
renderWindow.Render()

#------------Interactor--------------
iren = vtk.vtkRenderWindowInteractor()
iren.SetRenderWindow(renderWindow)
iren.Start()
```
this is the code for reading and i have the Cube.vtp in the same folder am i missing something?

**Reply by banesullivan** [reply]
Can you upload the saved file?

**Reply by banesullivan** [reply]
I notice a capital C in the first code and a lowercase in the second code for the file name... are you sure these match everywhere?

**Reply by abdelkareemkobo** [reply]
how do you want me to send the file?

**Reply by banesullivan** [reply]
Zip and attach to a comment here

**Reply by abdelkareemkobo** [reply]
[final_solution.zip](https://github.com/pyvista/pyvista/files/10667723/final_solution.zip)

**Reply by banesullivan** [reply]
You're mixing this up in a couple of ways. Please make sure the file name you pass to VTK matches exactly the file name you save out. 

In your code above, you pass `'cube.vtp'` to the reader. This file does not exist in the archive you attached. `'Cube.vtp'` with a capital-C does though. Make sure the filenames match.

Further, make sure you match the correct reader to the it's extension

`vtp` extension -> `vtkXMLPolyDataReader`
`vtk` extension -> `vtkDataSetReader`

**Reply by banesullivan** [reply]
Please note I made an update to use `vtkDataSetReader` over `vtkXMLReader` (as `vtkXMLReader` is abstract)

**Reply by abdelkareemkobo** [reply]
Okey , get it i will read the docs well and I am really happy with helping me , thanks my friend
